### DISCUSSION TITLE
Why we need to duplicate fields for different labels?

### DISCUSSION THREAD

**Discussion by IvanGoncharov** [graphql/defer-stream-wg#62] [General]
Following https://github.com/robrichard/defer-stream-wg/discussions/52#discussioncomment-4388444 by @yaacovCR it seems we adding a lot of complexity and workarounds to allow duplicating defer fields with different labels.

Given this query:
```graphql
{
  ... @defer(label: "a") {
    someField
  }
  ... @defer(label: "b") {
    someField
  }
  someField
}
```
we need to send value of `someField` 3 times.

Why do we need this?

# Solution:
I want to solve this separately from #52, so the solution is entirely based on the current proposal and doesn't include any ideas from #52.
I propose to treat incremental items as a path and follow the "merge patch" approach (an idea is borrowed from [RFC 7386](https://www.rfc-editor.org/rfc/rfc7386))
Given query:

```graphql
{
  ... @defer(label: "a") {
    f2
    f3
  }
  ... @defer(label: "b") {
    f2
    f4
  }
  f1
}
```

I suggest following response format:

```json
{
  "data": {
    "f1": 1
  },
  "hasNext": true
}
```

```json
{
  "incremental": [
    {
      "path": [],
      "data": {
        "f2": 2
      },
    }
  ],
  "hasNext": true
}
```

```json
{
  "incremental": [
    {
      "path": [],
      "fullfiledLabels": ["b"],
      "data": {
        "f4": 4
      },
    }
  ],
  "hasNext": true
}
```

```json
{
  "incremental": [
    {
      "path": [],
      "fullfiledLabels": ["a"],
      "data": {
        "f3": 3
      },
    }
  ],
  "hasNext": false
}
```

# Benefits

* Smaller payload
* Simple mental model of "partial response followed by optional series of patches" without any semantic overhead of response branching on "label"s
* Ability to defer individual fields without the overhead.

# Complex example1

```graphql
{
  ... @defer(label: "a") {
    f2 { a b c { d e f { h i } } }
  }
  ... @defer(label: "b") {
    f2 { a b c { d e f { h j } } }
  }
}
```

```json
{
  "data": {},
  "hasNext": true
}
```

```json
{
  "incremental": [
    {
      "path": [],
      "data": {
        "f2": {
          "a": "a",
          "b": "b",
          "c": {
            "d": "d",
            "e": "e",
            "f": [
              {
                "h": "h",
              }
            ]
          }
        }
      },
    }
  ],
  "hasNext": true
}
```

```json
{
  "incremental": [
    {
      "path": ["f2", "c", "f", 0],
      "fullfiledLabels": ["a"],
      "data": {
        "i": "i",
      },
    },
    {
      "path": [],
      "fullfiledLabels": ["a"],
    },
  ],
  "hasNext": true
}
```

```json
{
  "incremental": [
    {
      "path": ["f2", "c", "f", 0],
      "fullfiledLabels": ["a"],
      "data": {
        "j": "j",
      },
    },
    {
      "path": [],
      "fullfiledLabels": ["b"],
    },
  ],
  "hasNext": true
}
```

**Comment by glasser** [comment]
In your first example, it seems like the "state" between messages 2 and 3 (when we have f2 but not f3 or f4) does not correspond to any specific set of deferred fragments being resolved. That seems wrong to me? It seems like the whole point of putting defer at the fragment level is so that when doing codegen etc you can say "at any given time, either an entire deferred fragment is available or it's all missing (other than nested defers/streams/etc)". Which this breaks? I mean I think I like the overall concept, but I'd want to see f2 only come in with either f3 or f4.

**Comment by robrichard** [comment]
This proposal has evolved and is now being discussed here: https://github.com/robrichard/defer-stream-wg/discussions/65#discussioncomment-4757697
